### DISCUSSION TITLE
How to attach BagPart to a content item without migration?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#10929] [Q&A]
I am building a survey module where I use BagPart to attach many content-items to a single content item. Here is a summary of the overall design. 

 1. A `Survey` content type will have a BagPart named "Pages" attached to it using a migration. The "Survey.Pages" will contain multiple `Page` content type.
 2. A `SurveyPage` content type will have a BagPart named "Questions" attached to it using a migration. The "SurveyPage.Questions" will contain multiple `Question` content type.
 3. A `SurveyQuestion` content type will need to have a BagPart named "Inputs" attached to it NOT using migration. The "SurveyQuestion.Inputs" will contain multiple/different content-types where the stereotype is `SurveyInput`.

So in summary, a survey can have multiple pages. A page can have multiple questions. A question can have multiple inputs.

The reason I want to manually attach the "Inputs" BagPart to the SurveyQuestion is because I don't want the admin to be able to edit it using UI also I want to force which content-types can be attached to those with the `SurveyInput` stereotype. 

I tried to use a `ContentDisplayDriver` to attach BagPart so the `SurveyQuestion` like so

```C#
public class SurveyQuestionContentItemDisplayDriver : ContentDisplayDriver
{
    private readonly IContentDefinitionManager _contentDefinitionManager;

    public SurveyQuestionContentItemDisplayDriver(IContentDefinitionManager contentDefinitionManager)
    {
        _contentDefinitionManager = contentDefinitionManager;
    }

    public override bool CanHandleModel(ContentItem model)
    {
        return model.ContentType.Equals("SurveyQuestion");
    }

    public override Task<IDisplayResult> EditAsync(ContentItem model, BuildEditorContext context)
    {
        var part = new BagPart() 
       {
            ContentItem = model
       };
        model.Weld("Inputs", part);

        // attach the BagPart at the bottom of the Parts zone
        return Task.FromResult<IDisplayResult>(Initialize<BagPartEditViewModel>("BagPart_Edit", m =>
        {
            m.BagPart = part;
            m.Updater = context.Updater;
            m.ContainedContentTypeDefinitions = _contentDefinitionManager.ListTypeDefinitions().Where(t => t.GetSettings<ContentTypeSettings>().Stereotype == "SurveyControl");
        }).Location("Parts:100"));
    }

    public override async Task<IDisplayResult> UpdateAsync(ContentItem model, UpdateEditorContext context)
    {
        var part = model.Get<BagPart>("Inputs") ?? new BagPart();

        var vm = new BagPartEditViewModel { BagPart = part};

        await context.Updater.TryUpdateModelAsync(vm, Prefix);

        return Edit(model);
    }
}
```

However, the above code does not save the content items that were added to the bag! In other words, saving the Survey Content item saves everything except for the inputs. So when the Survey is loaded during the edit, no inputs will show up. The pages, question will get saved since these 2 were added using migration but the Inputs part does not get saved.

How can I save the Inputs data for the BagPart after it was manually attached?

Here is a screencast that'll show how the "Input" was not saved when everything else was

![description-issue-not-saving](https://user-images.githubusercontent.com/24724371/147595408-81b3356d-7f55-4e5c-a1be-b3432801b62a.gif)

**Comment by ns8482e** [comment]
Change `UpdateAsync` as following
```cs
 public override async Task<IDisplayResult> UpdateAsync(ContentItem model, UpdateEditorContext context)
    {
        await model.AlterAsync<BagPart>(model => updater.TryUpdateModelAsync(model, Prefix));
        return Edit(model);
    }
```

if that doesn't work try following

```cs
 public override async Task<IDisplayResult> UpdateAsync(ContentItem model, UpdateEditorContext context)
    {

        var vm = new BagPartEditViewModel() ;

        await context.Updater.TryUpdateModelAsync(vm, Prefix);

        var part =vm.BagPart  ?? new BagPart();

       model.Apply(part);

        return Edit(model);
    }
```

**Reply by MikeAlhayek** [reply]
Thanks for taking your time to look into this issue. Neither solution save the data. Its like the BagPartDisplayDriver never get called. I am wondering if I need to attach the BagPart using handler instead of driver.
